courses/views.py

```diff
@@ -20,13 +23,51 @@ class CourseViewSet(viewsets.ModelViewSet):
     lookup_field = 'uuid'
     serializer_class = CourseSerializer
 
+    def get_object(self):
+        """
+        Courses can be looked up via the uuid passed via kwargs (default
+        behavior) or through the course_id attribute in the payload.
+        """
+        queryset = self.get_queryset()
+
+        try:
+            obj = super(CourseViewSet, self).get_object()
+        except (AssertionError, Http404):
+            try:
+                obj = queryset.get(course_id=self.request.data.get('course_id'))
+            except Course.DoesNotExist:
+                pass
+        if not obj:
+            raise Http404(
+                "Could not find the course with uuid=%s or course_id=%s",
+                self.kwargs('uuid'), self.request.data.get('course_id')
+            )
+        return obj
+
     def create(self, request, *args, **kwargs):
         """
-        Prepopulate edx_instance based on the requesting user.
+        Incoming call from edX.
+
+        Because edX won't know if they've sent the information already or not,
+        this handles the update and create case.
+
+        It also does a bit of data preparation, setting the edx_instance from
+        the authenticated user and prefixing image paths with this
+        edx_instance.
         """
-        if not request.user.info.edx_instance:
-            raise ValidationError("User must have an associated edx_instance")
-        request.data['edx_instance'] = request.user.info.edx_instance
+        user = request.user
+        if not user.info.edx_instance:
+            raise ValidationError("User must have an associated edx_instance.")
+        request.data['edx_instance'] = user.info.edx_instance
+
+        if not request.data.get('image_url'):
+            raise ValidationError("You must specify an image_url")
+        request.data['image_url'] = parse.urljoin(
+            user.info.edx_instance, request.data['image_url'])
+
+        if Course.objects.filter(course_id=request.data['course_id']).exists():
+            return super(CourseViewSet, self).update(request, *args, **kwargs)
```

Comment:
It seems like we're bending over backward in order to make use of the super functions here and below. `get_object` looks for a uuid in two separate places. We're modifying `request.data` which may impact other uses of `request.data` elsewhere in the application. Can we pull the code we need from the super classes and modify it to our circumstances instead? Those mixins aren't many lines of code so I think the risk of not DRYing is minimal